### PR TITLE
feat: add a (secured) gateway endpoint for sending push notifications…

### PR DESCRIPTION
**Pull request by przemczan**
… from outside

### CODE REVIEW

File: src/controllers/gatewayController.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+router.post('/', async (request: Request, response: Response) => {
+    const message = request.body;
+    const validation = GatewayMessageSchema.validate(message);
+
+    if (validation.error) {
+        return sendResponse(response, false, validation.error.toString(), HttpStatus.BAD_REQUEST);
+    }
+
+    const subscriptions: SubscriptionModelType[] = await SubscriptionModel
```
Comment (kduralek) [inline src/controllers/gatewayController.ts]: No need to declare type here. Type inferring does the job.

File: src/server.ts [hunk 2]
```diff
@@ -6,51 +6,74 @@ import mongoose from 'mongoose';
...
+    config.webpush.mailto,
+    config.webpush.publicKey,
+    config.webpush.privateKey
+);
 
 const app: express.Application = express();
 
+const basic = auth.basic({
+        realm: 'RepuX WebPush'
+    }, function (username: string, password: string, callback: Function) {
```
Comment (kduralek) [inline src/server.ts]: arrow function for consitency

File: src/services/sendNotification.ts [hunk 3]
```diff
@@ -2,7 +2,7 @@ import logger from './../utils/logger';
 const webpush = require('web-push');
 
 export default async function sendNotification(data: any, params: any) {
```
Comment (kduralek) [inline src/services/sendNotification.ts]: why `async`? No `await` here

File: src/controllers/gatewayController.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        }) as SubscriptionModelType[];
+
+    if (subscriptions.length === 0) {
+        return sendResponse(response, false, 'No subscriptions found.');
+    }
+
+    try {
+        for (let subscription of subscriptions) {
+            const data = JSON.parse(subscription.body);
+            await sendNotification(data, message.params);
```
Comment (kduralek) [inline src/controllers/gatewayController.ts]: Do we really need to wait for each notification before return? Would be better to use promises and `Promise.all`? (parallel processing)

**Review by kduralek — APPROVED** [review]

**Comment by przemczan** [comment]
All suggestions taken into account.

**Pull request merged**